### ISSUE TITLE
[Feature] Change Provider trait

### ISSUE THREAD

**Issue by i1i1** [alloy-rs/alloy#568]
### Component

provider

### Describe the feature you would like

At this moment, `Provider` trait looks a bit weird:
https://github.com/alloy-rs/alloy/blob/cb95183d477024b57b27336035d10403e8ba55b8/crates/provider/src/provider.rs#L249-L253

There is always one to one correspondents between `Provider` and `Transport` and `Network` so if you plan on generalizing your structure over provider type you need to carry around both transport and network types in phantom data.

Check out this example:
```rust
struct MyProviderWrapper<P, T, N> {
    provider: P,
   _get_rid_of_me: PhantomData<(T, N)>,
}

impl<P, T, N> MyProviderWrapper<P, T, N>
where
    P: Provider<T, N>,
    T: Transport + Clone,
    N: Network,
{
    pub fn finally_we_can_use_our_transport(&self) { ... }
}
```

It would be great if at some moment it can be changed to look like this:
```rust
pub trait Provider {
    type Transport: Transport + Clone = BoxTransport;
    type Network: Network = Network;
    ...
}
```

So that, example would look like this:
```rust
struct MyProviderWrapper<P> {
    provider: P,
}

impl<P> MyProviderWrapper<P>
where
    P: Provider,
{
    pub fn finally_we_can_use_our_transport(&self) { ... }
}
```

Apart from that if you have some trait which relies on provider trait and needs to be generalized over transport it is not possible to make rust happy:
```rust
trait MyFancyTrait { ... }

// Not possible, beacuse T and N are not constrained anywhere
impl<P, T, N> MyFancyTrait for P
where
    P: Provider<T, N>,
    T: Transport + Clone,
    N: Network,
{
    ...
}
```

**Comment by onbjerg** [comment]
These can't be associated types simply because an associated type means that there is only one valid type for the trait implementation. The providers are generic over the network and transport because it works for any combination of network and transport.

If we made them associated types, you would have to implement the provider trait with pretty much the same implementation for every network and transport combination. This is not feasible, especially as the number of network implementations (e.g. arbitrum/optimism) and transport implementations grow

If you know what network and transport you are going to use in your code, you can make a type alias:

```rs
type MyProvider = Provider<Ethereum, BoxTransport>;
```

I hope that made sense, let me know what you think!

**Comment by onbjerg** [comment]
Ah, sorry, you're right. I am just waking up 🤦🏻  This is a good idea, I don't see any obvious downsides (other than this being breaking).

It stands to reason that for any provider with a single transport connected to some node, the network will be unchanged, so they are in fact always tied together in a 1:1 correspondence.

You are of course also correct that we can provide a blanket impl like your example.

Wonder what thoughts @prestwich / @mattsse have on this

**Comment by onbjerg** [comment]
Sorry, you're correct about the blanket implementation 🤦🏻 I'm just waking up. You are also correct about the 1:1 correspondence

As far as I know though, associated types are not object safe, correct? It is an invariant of ours that `Provider` is object safe

**Comment by prestwich** [comment]
Agree that current situation is not ideal ergonomics. Would love to remove phantomdata from provider layers, etc. 

The main reason that Provider is not already using assoc types is that [associated type defaults](https://rust-lang.github.io/rfcs/2532-associated-type-defaults.html) hasn't landed in the language yet. If it had, I would prettty happily set the assoc type defaults to `BoxTransport` and `Ethereum` and ship it. I'm reluctant to move to assoc types if it means giving up type defaults. 

My current belief about the API is that the "average user" should be using `Provider` with no generics to mean `Provider<BoxTransport, Ethereum>`, and that naming the generics should only be for code at a higher abstraction layer or with stricter needs than the marginal user. I'd basically like to push our examples towards not naming the `Transport` type  (which may need another typevar reordering from T,N to N,T so that we can name networks without also naming transports 😮‍💨)

> As far as I know though, associated types are not object safe, correct? It is an invariant of ours that Provider is object safe
> Edit: Actually, I think I'm incorrect here. As long as the associated types are not generic, we should be fine.

Edit is correct, yes. You can do `Box<dyn Provider<Transport = Http, Network = Ethereum>>` just fine. it's a bit more verbose, but not bad

**Comment by i1i1** [comment]
> The main reason that Provider is not already using assoc types is that [associated type defaults](https://rust-lang.github.io/rfcs/2532-associated-type-defaults.html) hasn't landed in the language yet. If it had, I would prettty happily set the assoc type defaults to BoxTransport and Ethereum and ship it. I'm reluctant to move to assoc types if it means giving up type defaults.

I have an ugly workaround for that. You can always duplicate trait, but without associated types:
```rust
trait ProviderWithDefaults {
    ...
}

impl<P: ProviderWithDefaults> Provider for P {
  type N = Ethereum;
  type T = BoxTransport;
  // Proxy requests to ProviderWithDefaults
}
```

> My current belief about the API is that the "average user" should be using Provider with no generics to mean Provider<BoxTransport, Ethereum>, and that naming the generics should only be for code at a higher abstraction layer or with stricter needs than the marginal user.

Let's break provider down. I think on average it makes sense to have a default for `Ethereum` as a network, but any transport can be boxed by default, right? So what is the point of restricting implementations only for Boxed version of it?

**Comment by prestwich** [comment]
i'll take a look at the associated type default workaround :)

**Comment by i1i1** [comment]
> I don't want to restrict implementations only to the box version. I want to ensure that the average user never sees the generic T at all. which can be achieved using the type defaults

I think there are 3 sides of this:
- There are users of trait which implement their own `Provider`. They would need either way to make transport and network explicit. There are no options for them
- There are users of trait which use it in trait bounds. By default, they shouldn't care about transport either way (it is implementation details) and generally they will rely on rpc calls
- People who either:
 + use `Box<dyn Provider<Transport = BoxTransport, Network = Ethereum>>`
 + use impl trait in return position like `fn f() -> impl Provider<Transport = BoxTransport, Network = Ethereum>`
I've grepped for both of these cases alloy and reth and I think `Provider` is just not used like that anywhere :shrug:

**Comment by prestwich** [comment]
I agree with you about the sides and the constraints it poses on downstream users from a technical perspective, but i think you're missing my point. I care about API experience for the most-common user, to the detriment of all other users. I do not really mind if it is difficult to write a `ProviderLayer` or a `TxFiller` as long as it is easy to write 

The user I care about is building a binary to make RPC calls to Ethereum (or maybe Optimism), and has a concrete provider type. This user wants the bound `P: Provider` with no additional qualifiers because their provider is a concrete type with `BoxTransport` and `Ethereum`. The only straightforward way to achieve that is with default types for `T` and `N`. Without default types, users will have to write `P: Provider<Transport = BoxTransport, Network = Ethereum>` or `provider: Box<dyn Provider<Transport = BoxTransport, Network = Ethereum>>` or something equally annoying for the average user to discover

The main target of this library is not people implementing their own `Provider` or their own layer or any other extension to the Provider system. It's not people  It's people making RPC calls to Ethereum

Basically I'm writing things for this user and am reluctant to approve anything that makes this bound more complex to write:

```
fn swap_something_on_uniswap<P: Provider>(
    provider: &P,
    token_in: Address,
    // ...
) { 
    // ...
}

**Comment by i1i1** [comment]
> The user I care about is building a binary to make RPC calls and send txns to Ethereum (or maybe Optimism). This user wants the bound P: Provider with no additional qualifiers because their provider is a concrete type with BoxTransport and Ethereum, instantiated with the ProviderBuilder during program init. The only straightforward way to achieve the simple P: Provider bound is with default types for T and N. Without default types, users will have to write P: Provider<Transport = BoxTransport, Network = Ethereum> or provider: Box<dyn Provider<Transport = BoxTransport, Network = Ethereum>> or something equally annoying for the average user to discover

Thanks for expanding on that!

I think generally you can still write `fn swap_on_uniswap<P: Provider>(provider: &P, token_in: Address)` with associated types. The only issues which might arise can be related to exact network type, but it will become an issue if you care about things which are received from RPC. I guess the only thing which user cares about are:
- constructing transaction (can be done via [TransactionBuilder trait](https://alloy-rs.github.io/alloy/alloy_network/trait.TransactionBuilder.html) or with generalized [CallBuilder](https://alloy-rs.github.io/alloy/alloy/contract/struct.CallBuilder.html#method.with_cloned_provider))
- getting events from contract
- Status of transactions

I think everything of these can be achieved without specifying concrete network type. WDUT?

**Comment by prestwich** [comment]
the average user cares a great deal about reading tx receipts, i.e. checking tx outcomes via logs. Receipts definitely needs a specified network type 😓

**Comment by i1i1** [comment]
> the average user cares a great deal about reading tx receipts, i.e. checking tx outcomes via logs. Receipts definitely needs a specified network type 😓

Yes! I think they can still watch logs as the logs are concrete types ([here](https://alloy-rs.github.io/alloy/alloy/providers/trait.Provider.html#method.get_logs)) and can be parsed later.

I know though that it would be handy to fetch logs like this:
```rust
provider.send_transaction().await?.get_receipt().await?.logs
```

But maybe we can add some fetchers for [receipt response](alloy::network::ReceiptResponse) apart from already existing ones. Like:
```rust
pub trait ReceiptResponse {
   ...
   fn logs() -> Option<Vec<alloy_primitive::Log>>;
}
```

What are your thoughts on this?

**Comment by prestwich** [comment]
```
pub trait ReceiptResponse {
   ...
   fn logs() -> Option<Vec<alloy_primitive::Log>>;
}
```

We have this as part of the `Receipt` trait currently, with the caveat that it is generic over `T`, for the following reasons:
- Logs may be parsed into `SolEvent` structs while remaining in the RpcLog and wrapping Receipt or RpcReceipt
- Other networks may choose to change their log type to contain additional information

**Comment by yash-atreya** [comment]
https://github.com/alloy-rs/alloy/pull/1859 somewhat addressed this.

**Comment by chris13524** [comment]
I've been very happy recently. I can just do `provider: &impl Provider` in my methods without any `where` shenanigans

**Comment by DaniPopes** [comment]
I think the main issue at hand here is UX of the Provider trait, which we thin we've improved significantly in the past few releases.

We removed the T generic and now the desired example would work fine:
```rust
struct MyProviderWrapper<P> {
    provider: P,
}

impl<P> MyProviderWrapper<P>
where
    P: Provider,
{
    pub fn finally_we_can_use_our_transport(&self) { ... }
}
```

You can also:
- keep the `N: Network` generic around if you're exporting this type in a library for maximum compatibility
- remove `P: Provider` and replace it with [`DynProvider`](https://docs.rs/alloy/latest/alloy/providers/struct.DynProvider.html) which is a thin wrapper around `Arc<dyn Provider<N>>`
